src/ResourceHelperTrait.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/*
+ * Copyright 2022 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore;
+
+use Google\ApiCore\ValidationException;
+
+trait ResourceHelperTrait
+{
+    private static $templateMap;
+
+    // Must be implemented by extendees to call loadPathTemplates.
+    private static function registerPathTemplates()
```

Comment:
The safety of having this marked abstract made a lot of sense. What are your thoughts on changing visibility to protected/public in order to achieve that? We can make sure we add annotations to the docblock about how it is internal/access should be considered private.

Reply:
> The safety of having this marked abstract made a lot of sense

Yeah, I liked a lot as well. It just _made sense_ to me as `abstract`.

> What are your thoughts on changing visibility to protected/public in order to achieve that? We can make sure we add annotations to the docblock about how it is internal/access should be considered private.

My only worry is that it really isn't something we want users to see on the surface. It's an implementation detail. The only `public` method on `GapicClientTrait` is `close` which makes a lot of sense. The `protected` functions have `@access private` phpdoc statements, but IDK why they need to be `protected` other than for maybe veneer testing purposes? The other thing was that since this `trait` should be used by generated surfaces only(?), we will always implement the `registerPathTemplates` function (he says with 100% confidence that nothing will ever go wrong with that assumption 😝).

If you & @bshaffer think it's nothing to worry about, I can switch it back to `protected abstract`.